cmake/modules/CheckCxxAtomic.cmake

```diff
@@ -10,18 +10,29 @@ function(check_cxx_atomics var)
     check_cxx_source_compiles("
 #include <atomic>
 #include <cstdint>
+
+#if __s390x__
+// Boost needs 16-byte atomics for tagged pointers.
+// These are implemented via inline instructions on the platform
+// if 16-byte alignment can be proven, and are delegated to libatomic
+// library routines otherwise.  Whether or not alignment is provably
+// OK for a std::atomic unfortunately depends on compiler version and
+// optimization levels, and also on the details of the expression.
+// We specifically test access via an otherwise unknown pointer here
+// to ensure we get the most complex case.  If this access can be
+// done without libatomic, then all accesses can be done.
+bool atomic16(std::atomic<unsigned __int128> *ptr)
```

Comment:
@uweigand can we call this method explicitly just to make sure it's not optimized out? something like
```c++
int main() {
  // ...
#if __s390x__
  std::atomic<unsigned __int128> w16;
  w1 = atomic16(&w16);
#endif
  return w1 + w2 + w4 + w8;
}
```

Reply:
I don't believe it **can** be optimized out, as it is a global function that could be provided by other modules.  I think we can assume the cmake tests will be built without LTO / whole-program assumptions ...

If it **could** be optimized out, then explicitly calling the method from main would not really help, because then the compiler could just inline it, and we'd we back at the problem of the current code.